content/interoperability.tex

```diff
@@ -0,0 +1,163 @@
+initialization call if the underlying software runtime of \openshmem and \ac{MPI}
+share the same internal communication resource.
+The program should always check the \VAR{provided} thread level returned
+at the corresponding initialization call or query the level of thread support
+after initialization to portably ensure thread support in each communication
+environment.
+
+Both \openshmem and \ac{MPI} define similar thread levels, namely, \VAR{THREAD\_SINGLE},
+\VAR{THREAD\_FUNNELED}, \VAR{THREAD\_SERIALIZED}, and \VAR{THREAD\_MULTIPLE}.
+When requesting threading support in a hybrid program, however,
+the following additional rules are applied if the implementations of \openshmem
+and \ac{MPI} share the same internal communication resource.
+It is strongly recommended to always follow these rules to ensure program
+portability.
+
+\begin{itemize}
+    \item The \VAR{THREAD\_SINGLE} thread level requires a single-threaded program.
+    Hence, a hybrid program should not request \VAR{THREAD\_SINGLE} at the initialization
+    call of either \openshmem or \ac{MPI} but request a different thread level at the
+    initialization call of the other model.
+
+    \item The \VAR{THREAD\_FUNNELED} thread level allows only the main thread to
+    make communication calls. A hybrid program using the \VAR{THREAD\_FUNNELED}
+    thread level in both \openshmem and \ac{MPI} should ensure that the same main thread
+    is used in both communication environments.
+
+    \item The \VAR{THREAD\_SERIALIZED} thread level requires the program to ensure
+    that communication calls are not made concurrently by multiple threads. If a
+    hybrid program uses \VAR{THREAD\_SERIALIZED} in one communication environment
+    and \VAR{THREAD\_SERIALIZED} or \VAR{THREAD\_FUNNELED} in the other one, it
+    should also guarantee that the \openshmem and \ac{MPI} calls are not made concurrently
+    from two distinct threads.
+\end{itemize}
+
+\subsection{Mapping Process Identification Numbers}
+\label{subsec:interoperability:id}
+
+Similar to the \ac{PE} number in \openshmem, \ac{MPI} defines rank as the
+identification number of a process in a communicator. Both the \openshmem \ac{PE}
+and the \ac{MPI} rank are unique integers assigned from zero to one less than the total
+number of processes. In a hybrid program, the \openshmem
+\ac{PE} number in \LibHandleRef{SHMEM\_TEAM\_WORLD}
+and the \ac{MPI} rank in \VAR{MPI\_COMM\_WORLD} of a process can be equal.
+This feature, however, may be provided by only some of the \openshmem and \ac{MPI}
+implementations (e.g., if both environments share the same underlying process
+manager) and is not portably guaranteed. A portable program should always
+use the standard functions in each model, namely, \FUNC{shmem\_team\_my\_pe} in \openshmem
+and \FUNC{MPI\_Comm\_rank} in \ac{MPI}, to query the process identification numbers
+in each communication environment and manage the mapping of identifiers in the
+program when necessary.
```

Comment:
Maybe it's implicit or maybe I'm missing something, but do we need to also specify that the size of `MPI_COMM_WORLD` is the same as `SHMEM_TEAM_WORLD`?  Otherwise, the example code below might not work correctly, right?  If we do need to specify the sizes are equal, do we have to also account for the usage of `MPI_Comm_Spawn`, which IIUC has a separate `MPI_COMM_WORLD` on the children ranks?

Reply:
Good question. You are right that there is no guarantee that size of `MPI_COMM_WORLD` is same as size of `SHMEM_TEAM_WORLD`. I think it depends on two factors:
1.  How the program initializes SHMEM and MPI (e.g., some processes call `MPI_Init` but do not call `shmem_init` )
2. How the implementations define accessible processes (e.g., although all launched processes call both SHMEM and MPI init, some of them might be visible in MPI but not visible in SHMEM).

For the example code, it forces each process to initialize both MPI and SHMEM, thus factor-(1) is confirmed. For factor-(2), I think only the processes who are accessible in both models can execute the code (e.g., any hidden process should not be able to execute the code after init calls). Thus, the size of each "WORLD" should be equal and the program works correctly.

For the program using `MPI_Comm_Spawn`, it should not use SHMEM to communicate with dynamically created processes. We have explicitly defined it in section "Dynamic Process Creation".

Please let me know if I miss any case.

Reply:
Yes, I think you're correct - thanks!